### PR TITLE
add design proposal template

### PR DESCRIPTION
**Pull request by xmulligan**
fixing #35

### CODE REVIEW

**Comment by carolynvs** [comment]
Sorry I messed up the first PR (#35). I have flagged this for review from the governance WG and our TOC liaisons. @jberkus or @geekygirldawn  Can you ping our liaisons for review?

**Comment by geekygirldawn** [comment]
Sorry about this - I think it fell through the cracks in the run up to KubeCon :(

I'm thinking that maybe it shouldn't be named GOVERNANCE-design-proposals.md, but maybe instead called DESIGN-PROPOSALS.md, since it's not really part of the GOVERNANCE.md file. The other files called GOVERNANCE-something.md
are meant to be customized and used as GOVERNANCE.md.

@jberkus did you have any thoughts before this goes to our liaisons for approval?

Also, I know that @jberkus is at a conference this week, so we may need to catch up with him next week for this one.

**Comment by xmulligan** [comment]
No worries, totally understand that! I've updated the file name and will wait for @jberkus to comment when he is back from the conference

File: DESIGN-PROPOSALS.md
Comment (jberkus) [inline DESIGN-PROPOSALS.md:7]:
```suggestion
**(optional) Status: **implementable, accepted, deferred, rejected, withdrawn, or replaced

**Checklist**:

- [ ] Prerequisite 1
- [ ] Prerequisite 2
- [ ] Component 1
- [ ] Component 2
- [ ] Docs
- [ ] Tests


```

We should have a checklist in here so that people can track how much of the proposal has been implemented.  It's part of the KEP template.

**Review by jberkus — APPROVED** [review]

**Comment by jberkus** [comment]
Looks good, I'll ping our CNCF liasons.

File: DESIGN-PROPOSALS.md
Comment (TheFoxAtWork) [inline DESIGN-PROPOSALS.md:36]:
```suggestion
Detail the things that people will be able to do with the feature if it is implemented. Include as much detail as possible so that people can understand the "how" of the system. This can also be combined with the prior sections.
```

File: DESIGN-PROPOSALS.md
Comment (TheFoxAtWork) [inline DESIGN-PROPOSALS.md:40]:
```suggestion
List the desired goal or goals that the design is intended to achieve. These goals can be leveraged to structure and scope the design and discussions and may be reused as the "definition of done" -  the criteria you use to know the implementation of the design has succeeded in accomplishing its goals.
```

File: DESIGN-PROPOSALS.md
Comment (TheFoxAtWork) [inline DESIGN-PROPOSALS.md:48]:
```suggestion
This is where we get down to the specifics of what the proposal actually is. It should have enough detail that reviewers can understand exactly what you're proposing, but should not include things like API designs or implementation. This section should expand on the desired outcome and include details on how to measure success.
```

File: DESIGN-PROPOSALS.md
Comment (TheFoxAtWork) [inline DESIGN-PROPOSALS.md:52]:
```suggestion
This section should contain enough information that the specifics of the change are understandable. This may include API specs (though not always required) or even code snippets. If there's any ambiguity about HOW your proposal will be implemented, this is the place to discuss them. This can also be combined with the proposal section above. It should also address how the solution is backward compatible and how to deal with these incompatibilities, possibly with defaulting or migrations. It may be useful to refer back to the goals and non-goals to assist in articulating the "why" behind your approach.
```

File: DESIGN-PROPOSALS.md
Comment (TheFoxAtWork) [inline DESIGN-PROPOSALS.md:63]:
```suggestion
Cons are defined as the negative aspects or disadvantages of the design as described. This section has the potential to capture outstanding challenge areas or future improvements needed for the project and could be referenced in future PRs and issues. This is also a good place to check for any assumptions that have been made in the design.
```

File: DESIGN-PROPOSALS.md
Comment (TheFoxAtWork) [inline DESIGN-PROPOSALS.md:66]:
```suggestion
Describe the risks of this proposal and how they can be mitigated. This should be broadly scoped and describe how it will impact the larger ecosystem and potentially adopters of the project; such as if adopters need to immediately update, or support a new port or protocol. It should include drawbacks to the proposed solution and consider the security implications of the changes too. 

When attempting to identify security implications of the changes, consider the following questions:
* Does the change alter the permissions or access of users, services, components - this could be an improvement or downgrade or even just a different way of doing it?
* Does the change alter the flow of information, events, and logs stored, processed, or transmitted?
* Does the change increase the 'surface area' exposed - meaning, if an operator of the project or user were to go rogue or be uninformed in its operation, do they have more areas that could be manipulated unfavorably?
* What existing security features, controls, or boundaries would be affected by this change? ```

File: DESIGN-PROPOSALS.md
Comment (TheFoxAtWork) [inline DESIGN-PROPOSALS.md:93]:
```suggestion
Describe the development and implementation phases planned to break up the work and/or record them here as they occur. Provide enough detail so readers may track the major milestones in the lifecycle of the design proposal and correlate them with issues, PRs, and releases occurring within the project.
```

**Review by TheFoxAtWork — COMMENTED** [review]
I've provided a lot of comments, some are pure suggestions, some are proposed rewrites to improve clarity or robustness of the desired content - they're not required as i want to ensure i understand the intent in those areas and leave that to the authors to confirm, clarify, or reject.

File: DESIGN-PROPOSALS.md
Comment (dzolotusky) [inline DESIGN-PROPOSALS.md:16]: Should these "prerequisite" and "component" lines be something else?
I read @jberkus's comment to suggest that this checklist be all of the phases in the lifecycle of a design proposal, and they get checked off as the proposal moves through those phases. I assume that "prerequisite 1"... "Component 2" aren't phases in the lifecycle.

File: DESIGN-PROPOSALS.md
Comment (dzolotusky) [reply DESIGN-PROPOSALS.md:36]: I'd also add something about who the target users that will be using this feature are.

File: DESIGN-PROPOSALS.md
Comment (dzolotusky) [reply DESIGN-PROPOSALS.md:16]: If the idea is for the authors to fill this checklist with pieces of the proposal and not lifecycle stages of a proposal, I think that we need to make that clear with a sentence or two above the checklist explaining how an author will use it.

File: DESIGN-PROPOSALS.md
Comment (dzolotusky) [reply DESIGN-PROPOSALS.md:24]: "recording change notes for the project" doesn't sound like what this is. I think that we need to clarify that this is for template for creating a design change or feature in a project and for tracking that as it gets implemented

File: DESIGN-PROPOSALS.md
Comment (dzolotusky) [reply DESIGN-PROPOSALS.md:52]: I'd change "specifics" to something more, umm, specific. Like, should the section contain enough informaiton that the "proposed implementation" can be understood? Is this still for more depth to the product direction and not tech/implementation details?

File: DESIGN-PROPOSALS.md
Comment (dzolotusky) [reply DESIGN-PROPOSALS.md:66]: let's create a new section for security

File: DESIGN-PROPOSALS.md
Comment (TheFoxAtWork) [inline DESIGN-PROPOSALS.md:16]:
```suggestion
- [ ] Problem, need, impact, and solution definition
- [ ] User Story/Stories
- [ ] Proposal scoping (goals/non-goals)
- [ ] Detailed design, diagrams
- [ ] Impact, risk, and security considerations
- [ ] Observability, performance, and scaling considerations
- [ ] Availability, compatibility, and roll back considerations
- [ ] Implementation phases, planning timeline
- [ ] Test plan
- [ ] Accept/reject proposal
- [ ] Implementation
- [ ] Docs updates and release notes
- [ ] Release
```

I've suggested a more specific listing that aligns with the content of this doc. If this is more what you are looking for, i would recommend shifting the ordering of the doc to reflect the list here so authors can follow along (it'd be sweet if these were links to the sections of the doc, then it can be used as a Table of Contents )

**Review by TheFoxAtWork — COMMENTED** [review]
adjusted my suggestions based on @dzolotusky 's comments.  Dave, please take a second look to see if the suggestions would resolve your comments. @jberkus @xmulligan please also review as this is based on our understanding of the content presented, which may not align with the original intent.

File: DESIGN-PROPOSALS.md
Comment (jberkus) [reply DESIGN-PROPOSALS.md:16]: Well, I was mostly trying to indicate that there should be some kind of checklist there, but what goes *in* the checklist is highly dependent on the project's own processes.  For example, KEPs have a very different-looking checklist: https://github.com/kubernetes/enhancements/issues/3668 .  I was trying NOT to have anything that suggested a specific process that probably wouldn't apply.

File: DESIGN-PROPOSALS.md
Comment (jberkus) [reply DESIGN-PROPOSALS.md:16]: It's supposed to be a generic "There should be a checklist here".  What goes in the checklist is highly dependent on how the individual project manages their PRs.

But regardless, I think we're getting to the point of "we can't move foward on this without also having documentation for the template user".

**Comment by jberkus** [comment]
@xmulligan based on the feedback above, I think this template is going to need an accompanying documentation to be useful. Was that something you'd started on, or do we need to recruit someone else for that?

File: DESIGN-PROPOSALS.md
Comment (xmulligan) [inline DESIGN-PROPOSALS.md:93]:
```suggestion
* What existing security features, controls, or boundaries would be affected by this change?
```

File: DESIGN-PROPOSALS.md
Comment (xmulligan) [reply DESIGN-PROPOSALS.md:16]: I agree with Josh that this should be a checklist and not a table of contents.

Comment (TheFoxAtWork) [reply DESIGN-PROPOSALS.md:16]: Sounds good.  Providing instructions for this checklist would be helpful, particularly describing what it is intended to capture.

**Review by TheFoxAtWork — COMMENTED** [review]
Just one explanation on the checklist and this will be good to go for me

**Review by TheFoxAtWork — APPROVED** [review]

**Comment by jberkus** [comment]
@dzolotusky @TheFoxAtWork is this good enough now?

**Comment by TheFoxAtWork** [comment]
You've got my approval.

**Comment by geekygirldawn** [comment]
@dzolotusky - are you with these changes?

**Comment by dzolotusky** [comment]
> @dzolotusky - are you with these changes?

Yes, this looks good to me now.

Sorry for taking so long to get back to you here.

**Review by dzolotusky — APPROVED** [review]

**Comment by geekygirldawn** [comment]
@amye @TheFoxAtWork - this PR is done, so if one of you could merge it, that would be great!

**Comment by TheFoxAtWork** [comment]
@geekygirldawn , GitHub is showing you have an outstanding action to review and approve this pull request. I am also not authorized to merge it.  Once you review and approve, @amye should be able to merge.

**Review by geekygirldawn — APPROVED** [review]
Thanks for all of your hard work on this!

**Pull request merged**
